src/contrib/tevent.rst

```diff
@@ -0,0 +1,655 @@
+        if (req == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "tevent_req_create() failed\n");
+            return NULL;
+        }
+
+        subreq = query_ldap_send(state, ev, "(name=%s)", name);
+        if (subreq == NULL) {
+            ret = ENOMEM;
+            goto done;
+        }
+
+        tevent_req_set_callback(subreq, fetch_user_done, req);
+
+        return req;
+
+    done:
+        tevent_req_error(req, ret);
+        tevent_req_post(req, ev);
+
+        return req;
+    }
+
+    static void fetch_user_done(struct tevent_req *subreq)
+    {
+        struct fetch_user_state *state;
+        struct tevent_req *req;
+        errno_t ret;
+
+        req = tevent_req_callback_data(subreq, struct tevent_req);
+        state = tevent_req_data(req, struct fetch_user_state);
+
+        ret = query_ldap_recv(state, subreq, &state->result);
+        talloc_zfree(subreq);
+        if (ret != EOK) {
+            goto done;
+        }
+
+    done:
+        if (ret != EOK) {
+            tevent_req_error(req, ret);
+            return;
+        }
+
+        tevent_req_done(req);
+    }
+
+    errno_t fetch_user_recv(TALLOC_CTX *mem_ctx,
+                            struct tevent_req *req,
+                            struct ldap_result **_result)
+    {
+        struct fetch_user_state *state = NULL;
+        state = tevent_req_data(req, struct fetch_user_state);
+
+        TEVENT_REQ_RETURN_ON_ERROR(req);
+
+        *_result = talloc_steal(mem_ctx, state->result);
+
+        return EOK;
+    }
+
+.. note::
+
+    If an error occurs before the request had a chance to step into an
+    asynchronous processing (that is any error in the send function) before the
+    caller had a chance to register a callback, it is necessary to call
+    ``tevent_req_post()``. This makes sure that the callback is executed once it
+    is registered.
+
+    If the request finished successfully it is terminated with
+    ``tevent_req_done``. If it ends with an error then use ``tevent_req_error``.
+    Both functions will execute the caller's callback immediately if it is
+    available.
+
+
+Calling the request
+===================
+
+The previous chapter showed how the request is created. The following example
+shows how it is used.
+
+.. code-block:: c
+
+    errno_t caller(TALLOC_CTX *mem_ctx,
+                struct tevent_context *ev)
+    {
+        struct tevent_req *req;
+        errno_t ret;
+
+        req = fetch_user_send(mem_Ctx, ev, "John Doe");
+        if (req == NULL) {
+            return ENOMEM;
+        }
+
+        tevent_req_set_callback(req, caller_done, /* private data if any */ NULL);
+    }
+
+    void caller_done(struct tevent_req *req)
+    {
+        struct ldap_result *result;
+
+        ret = fetch_user_recv(NULL, subreq, &result);
+        talloc_zfree(subreq);
+        if (ret != EOK) {
+            goto done;
+        }
+
+        print_result(result);
+    }
+
```

Comment:
I think having here diagram from line 384 populated with function names used in above code examples could be nice.

Reply:
Good idea. Done.